lib/Sema/DerivedConformance/DerivedConformanceDistributedActor.cpp

```diff
@@ -332,17 +332,24 @@ deriveBodyDistributed_invokeHandlerOnReturn(AbstractFunctionDecl *afd,
     metatypeVar->setImplicit();
     metatypeVar->setSynthesized();
 
+    // If the SerializationRequirement requires it, we need to emit a cast:
     // metatype as! <<concrete SerializationRequirement.Type>>
+    bool serializationRequirementIsAny =
+        metatypeParam->getInterfaceType()->getMetatypeInstanceType()->isEqual(
+            C.getAnyExistentialType());
+
     auto metatypeRef =
         new (C) DeclRefExpr(ConcreteDeclRef(metatypeParam), dloc, implicit);
-    auto metatypeSRCastExpr = ForcedCheckedCastExpr::createImplicit(
-        C, metatypeRef, serializationRequirementMetaTypeTy);
-
     auto metatypePattern = NamedPattern::createImplicit(C, metatypeVar);
-    auto metatypePB = PatternBindingDecl::createImplicit(
-        C, swift::StaticSpellingKind::None, metatypePattern,
-        /*expr=*/metatypeSRCastExpr, func);
 
+    PatternBindingDecl *metatypePB = serializationRequirementIsAny ?
+      PatternBindingDecl::createImplicit(
+          C, swift::StaticSpellingKind::None, metatypePattern,
+          /*expr=*/metatypeRef, func) :
+      PatternBindingDecl::createImplicit(
+          C, swift::StaticSpellingKind::None, metatypePattern,
+          /*expr=*/ForcedCheckedCastExpr::createImplicit(
+          C, metatypeRef, serializationRequirementMetaTypeTy), func);
```

Comment:
This basically just avoids the cast `as! Any` because we'd get a crash later on in trying to typecheck and diagnose the un-necessary cast because the castTypeRepr we'd get in

```
auto castTypeRepr = expr->getCastTypeRepr();
      emitDiagnostic(diag::forced_downcast_noop, toType)
          .fixItRemove(SourceRange(diagLoc, castTypeRepr->getSourceRange().End));
```

would be null... I guess avoiding the cast entirely might be good to do; I couldn't quite figure out why the type there would ever become null, as we clearly do pass the `metatypePattern` here everywhere hm

Reply:
Yeah, the cast is there because we're getting some `R` that is unconstrained and then we need to cast it to the serialization requirement as we pass it into user functions that require `<R: TheSerializationRequirement>`:


```
<unknown>:0: error: local function 'doInvokeOnReturn(returnType:)' requires that 'R' conform to 'Encodable'
<unknown>:0: error: local function 'doInvokeOnReturn(returnType:)' requires that 'R' conform to 'Decodable'
```

Reply:
I was unsure if I want to fix the diagnostic but that's a good point, let me include that here

Reply:
Yeah, good point. I originally though to fix the warning but then moved to fix the problem; but BOTH fixes actually make sense... Let me do that as well [3595b34b6d7a833b9f6f561c5a6d67135ec0c83e](https://github.com/swiftlang/swift/pull/84119/commits/3595b34b6d7a833b9f6f561c5a6d67135ec0c83e)